subsys/fs/littlefs_fs.c

```diff
@@ -0,0 +1,663 @@
+
+	fs_unlock(fs);
+	return lfs_to_errno(ret);
+}
+
+static ssize_t littlefs_read(struct fs_file_t *fp, void *ptr, size_t len)
+{
+	struct fs_littlefs *fs = fp->mp->fs_data;
+
+	fs_lock(fs);
+
+	ssize_t ret = lfs_file_read(&fs->lfs, LFS_FILEP(fp), ptr, len);
+
+	fs_unlock(fs);
+	return lfs_to_errno(ret);
+}
+
+static ssize_t littlefs_write(struct fs_file_t *fp, const void *ptr, size_t len)
+{
+	struct fs_littlefs *fs = fp->mp->fs_data;
+
+	fs_lock(fs);
+
+	ssize_t ret = lfs_file_write(&fs->lfs, LFS_FILEP(fp), ptr, len);
+
+	fs_unlock(fs);
+	return lfs_to_errno(ret);
+}
+
+BUILD_ASSERT((FS_SEEK_SET == LFS_SEEK_SET)
+	     && (FS_SEEK_CUR == LFS_SEEK_CUR)
+	     && (FS_SEEK_END == LFS_SEEK_END));
+
+static int littlefs_seek(struct fs_file_t *fp, off_t off, int whence)
+{
+	struct fs_littlefs *fs = fp->mp->fs_data;
+
+	fs_lock(fs);
+
+	off_t ret = lfs_file_seek(&fs->lfs, LFS_FILEP(fp), off, whence);
+
+	fs_unlock(fs);
+
+	if (ret >= 0) {
+		ret = 0;
+	}
+
+	return lfs_to_errno(ret);
+}
+
+static off_t littlefs_tell(struct fs_file_t *fp)
+{
+	struct fs_littlefs *fs = fp->mp->fs_data;
+
+	fs_lock(fs);
+
+	off_t ret = lfs_file_tell(&fs->lfs, LFS_FILEP(fp));
+
+	fs_unlock(fs);
+	return ret;
+}
+
+static int littlefs_truncate(struct fs_file_t *fp, off_t length)
+{
+	struct fs_littlefs *fs = fp->mp->fs_data;
+
+	fs_lock(fs);
+
+	int ret = lfs_file_truncate(&fs->lfs, LFS_FILEP(fp), length);
+
+	fs_unlock(fs);
+	return lfs_to_errno(ret);
+}
+
+static int littlefs_sync(struct fs_file_t *fp)
+{
+	struct fs_littlefs *fs = fp->mp->fs_data;
+
+	fs_lock(fs);
+
+	int ret = lfs_file_sync(&fs->lfs, LFS_FILEP(fp));
+
+	fs_unlock(fs);
+	return lfs_to_errno(ret);
+}
+
+static int littlefs_mkdir(struct fs_mount_t *mountp, const char *path)
+{
+	struct fs_littlefs *fs = mountp->fs_data;
+
+	path = fs_impl_strip_prefix(path, mountp);
+	fs_lock(fs);
+
+	int ret = lfs_mkdir(&fs->lfs, path);
+
+	fs_unlock(fs);
+	return lfs_to_errno(ret);
+}
+
+static int littlefs_opendir(struct fs_dir_t *dp, const char *path)
+{
+	struct fs_littlefs *fs = dp->mp->fs_data;
+
+	if (k_mem_slab_alloc(&lfs_dir_pool, &dp->dirp, K_NO_WAIT) != 0) {
+		return -ENOMEM;
+	}
+
+	memset(dp->dirp, 0, sizeof(struct lfs_dir));
+
+	path = fs_impl_strip_prefix(path, dp->mp);
+
+	fs_lock(fs);
+
+	int ret = lfs_dir_open(&fs->lfs, dp->dirp, path);
+
+	fs_unlock(fs);
+
+	if (ret < 0) {
+		k_mem_slab_free(&lfs_dir_pool, &dp->dirp);
+	}
+
+	return lfs_to_errno(ret);
+}
+
+static void info_to_dirent(const struct lfs_info *info, struct fs_dirent *entry)
+{
+	entry->type = ((info->type == LFS_TYPE_DIR) ?
+		       FS_DIR_ENTRY_DIR : FS_DIR_ENTRY_FILE);
+	entry->size = info->size;
+	strncpy(entry->name, info->name, sizeof(entry->name));
+	entry->name[sizeof(entry->name) - 1] = '\0';
+}
+
+static int littlefs_readdir(struct fs_dir_t *dp, struct fs_dirent *entry)
+{
+	struct fs_littlefs *fs = dp->mp->fs_data;
+
+	fs_lock(fs);
+
+	struct lfs_info info;
+	int ret = lfs_dir_read(&fs->lfs, dp->dirp, &info);
```

Comment:
I've just noticed that littlefs add by default '.' and '..' directory entries (and it's not configurable in LTS module). This is opposite to FatFS and NFFS behaviour. We can remove them here or leave decisions on higher layers.

Reply:
Yeah.  I'm pretty ambivalent about it.  I don't know about NFFS, but I don't think FatFS generates those entries (later: apparently it can); if neither of their wrappers are filtering the results, there's an argument that littlefs shouldn't either.

Here's the upstream rationale for their presence: https://github.com/ARMmbed/littlefs/issues/175#issuecomment-493103440

On the other hand the entries really aren't very useful (they're always the first two returned from readdir).  If existing users of the Zephyr FS API are not prepared to handle them, it'd be better to remove them at the wrapper layer (and document in the API that they are not to be provided by implementations).

Absent other comment I'll go that route.  Thanks for bringing this up.

Reply:
>  better to remove them at the wrapper layer (and document in the API that they are not to be provided by implementations).

Except I just realized that this file system may be used under the POSIX wrappers, where they would be expected and handled properly.   @jimparis @pfalcon any comment?

Whichever way this goes the rationale needs to be documented:
* `fs_readdir` returns all entries supplied by the underlying file system implementation; or
* `fs_readdir` strips `"."` and `".."` for consistency with other file system implementations.

Reply:
After looking at it I think the preferred solution is outside the scope of littlefs, so I'm punting this to #17951, and will keep the current approach of not filtering littlefs results.

Reply:
@pabigot keep in mind that the FUSE fs access layer is adding them to be posix compliant.

https://github.com/zephyrproject-rtos/zephyr/blob/af9a2670d254a865faa6bff4261995616a23f48a/subsys/fs/fuse_fs_access.c#L203-L204

Reply:
Sounds like FUSE is doing that because the FS abstraction isn't well defined, and option 2 of #17951 is the right way to go.